doc/how-to-source-packages.rst

```diff
@@ -0,0 +1,309 @@
+
+- For an exact version of the package, use the package name and version.
+
+   .. code-block:: shell
+
+      $ VER=0.12.3.1 cabal get vector-{$VER}
+      Unpacking to vector-0.12.3.1/
+      $ VER=0.12.3.1 echo "packages: vector-$VER" >> cabal.project
+
+- For the latest version of the package, use the package name only after a ``cabal update``.
+
+   .. code-block:: shell
+
+      $ cabal update
+      $ cabal get vector
+      Unpacking to vector-0.13.1.0/
+      $ echo "packages: vector-0.13.1.0" >> cabal.project
+
+You can vendor unpacked source code obtained by other means, by means that
+replace the ``cabal unpack`` step, in the same way.
+
+Fork don't vendor
+^^^^^^^^^^^^^^^^^
+There's no need to vendor packages on Hackage if you expect Hackage to always be
+available as packages cannot be deleted from Hackage. Source code repositories,
+on the other hand, can disappear.
+
+Rather than vendoring, it might be easier to take a
+``source-repository-package`` dependency on a fork of the upstream source code
+repository, on a fork that you control, especially if you're going to be making
+contributions to the upstream repository.
+
+.. _pkg-authors:
+
+*Source code* as a package author or maintainer
+-----------------------------------------------
+
+If you are authoring or maintaining a package and want to link to the
+:ref:`package source<pkg-author-source>` you will be dealing with
+``source-repository``, a package description stanza that typically specifies
+where upstream development of the package is happening.  A package published to
+Hackage will be displayed with a link to the source repository if a
+``source-repository`` stanza is present in the package description.
+
+This helps users find the source code for the package if they want to contribute
+to it or if they want to try out the latest changes before they're published.
+The simplest way to try out the latest changes is to add a
+``source-repository-package`` dependency on the package so that it will be
+picked up from the source code repository instead of from Hackage.
+
+.. Note::
+
+   It is easy to copy the fields of ``source-repository`` to set up a
+   ``source-repository-package`` project source code repository dependency as
+   they share many of the same :ref:`VCS fields<vcs-fields>`. Let's see an
+   example, again using the ``vector`` package;
+
+   Looking at vector's `package description <vector-pkg-desc_>`_ on Hackage we
+   see the ``source-repository`` stanza:
+
+   .. code-block:: cabal
+
+      source-repository head
+         type:     git
+         location: https://github.com/haskell/vector.git
+         subdir:   vector
+
+   To turn that into a dependency we'd add the following to our project, adding
+   a ``tag`` field if we wanted to pin to a particular commit;
+
+   .. code-block:: cabal
+
+      source-repository-package
+         type:     git
+         location: https://github.com/haskell/vector.git
+         subdir:   vector
+
```

Comment:
It would be good to add a tag field to the example to point to the 0.13.1.0 version and continue the theme of using that version throughout the  guide page

Reply:
Can do now that I see there's a tag for it. This version is not in [github/vector releases](https://github.com/haskell/vector/releases) but there is a [vector-0.13.1.0 tag](https://github.com/haskell/vector/tree/vector-0.13.1.0).

Reply:
I've include the tag field in the example and shown how I got that commit hash, with;

```
$ git ls-remote --tags
From git@github.com:haskell/vector.git
...
79bdd2edcfaf6b07f7fabc43a7d9c5a2ff93d3ca     refs/tags/vector-0.13.1.0
```

I also moved this content, that was in a note, up to a subsection. I gave it the title of "Setting up a source code dependency".